src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,712 @@
+        public bool TryAdd(TKey key, TValue value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_arrayStorage != null)
+            {
+                if (ContainsKeyCore(key))
+                {
+                    return false;
+                }
+
+                EnsureCapacity(Count + 1);
+
+                if (_dictionaryStorage != null)
+                {
+                    return _dictionaryStorage.TryAdd(key, value);
+                }
+
+                _arrayStorage[Count] = new KeyValuePair<TKey, TValue>(key, value);
+                _count++;
+                return true;
+            }
+
+            return _dictionaryStorage!.TryAdd(key, value);
+        }
+
+        /// <inheritdoc />
+        public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_arrayStorage != null)
+            {
+                return TryFindItem(key, out value);
+            }
+
+            return _dictionaryStorage!.TryGetValue(key, out value);
+        }
+
+        [DoesNotReturn]
+        private static void ThrowArgumentNullExceptionForKey()
+        {
+            throw new ArgumentNullException("key");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnsureCapacity(int capacity)
+        {
+            EnsureCapacitySlow(capacity);
+        }
+
+        private void EnsureCapacitySlow(int capacity)
+        {
+            Debug.Assert(_arrayStorage != null);
+
+            if (_arrayStorage.Length < capacity)
+            {
+                if (capacity > DefaultArrayThreshold)
+                {
+                    _dictionaryStorage = new Dictionary<TKey, TValue>(capacity);
+                    foreach (var item in _arrayStorage)
+                    {
+                        _dictionaryStorage[item.Key] = item.Value;
+                    }
+
+                    // Clear array storage.
+                    _arrayStorage = null;
+                }
+                else
+                {
+                    capacity = _arrayStorage.Length == 0 ? DefaultArrayThreshold : _arrayStorage.Length * 2;
+                    var array = new KeyValuePair<TKey, TValue>[capacity];
+                    if (_count > 0)
+                    {
+                        Array.Copy(_arrayStorage, 0, array, 0, _count);
+                    }
+
+                    _arrayStorage = array;
+                }
+            }
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int FindIndex(TKey key)
+        {
+            // Generally the bounds checking here will be elided by the JIT because this will be called
```

Comment:
This comment is outdated, and maybe not true either.
E.g. `Remove` calls this method, but doesn't call `EnsureCapacity`.